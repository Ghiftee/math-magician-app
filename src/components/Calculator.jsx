import React, { useState } from 'react';
import calculate from '../logic/calculate';
import '../logic/operate';

export default function Calculator() {
  const [calc, setValues] = useState({
    operation: null,
    next: null,
    total: null,
    error: null,
  });
  const buttonClicked = (e) => {
    const { error } = calc;
    if (error) {
      setValues({ ...calc, error: null });
    }
    const buttonName = e.target.innerText || e.target.innerHTML;
    let result;
    try {
      result = calculate(calc, buttonName);
    } catch (err) {
      setValues({
        operation: null,
        next: null,
        total: 0,
        error: err,
      });
      return;
    }

    const { operation, next, total } = calc;
    setValues({
      operation: ((result.operation === undefined) ? operation : result.operation),
      next: ((result.next === undefined) ? next : result.next),
      total: ((result.total === undefined) ? total : result.total),
    });
  };
  const { next, total, error } = calc;
  return (
    <div id="calculator">
      <input className="text-right border-0 bg-gray text-white p-1 py-3 w-100" value={error || next || (total || 0)} readOnly />
      <div className="buttons d-grid bg-gray">
        <button type="button" onClick={buttonClicked}>AC</button>
        <button type="button" onClick={buttonClicked}>+/-</button>
        <button type="button" onClick={buttonClicked}>%</button>
        <button className="bg-warning" type="button" onClick={buttonClicked}>รท</button>
        <button type="button" onClick={buttonClicked}>7</button>
        <button type="button" onClick={buttonClicked}>8</button>
        <button type="button" onClick={buttonClicked}>9</button>
        <button className="bg-warning" type="button" onClick={buttonClicked}>x</button>
        <button type="button" onClick={buttonClicked}>4</button>
        <button type="button" onClick={buttonClicked}>5</button>
        <button type="button" onClick={buttonClicked}>6</button>
        <button className="bg-warning" type="button" onClick={buttonClicked}>-</button>
        <button type="button" onClick={buttonClicked}>1</button>
        <button type="button" onClick={buttonClicked}>2</button>
        <button type="button" onClick={buttonClicked}>3</button>
        <button className="bg-warning" type="button" onClick={buttonClicked}>+</button>
        <button id="zero-button" type="button" onClick={buttonClicked}>0</button>
        <button type="button" onClick={buttonClicked}>.</button>
        <button className="bg-warning" type="button" onClick={buttonClicked}>=</button>
      </div>
    </div>
  );
}
