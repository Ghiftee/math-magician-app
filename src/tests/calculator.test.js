import { render, screen, fireEvent } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import Calculator from '../components/Calculator';
import CalcPage from '../pages/calculator';
import Home from '../pages/home';

test('expect calculator element to match working snapshot', () => {
  const calculatorComponent = renderer.create(<Calculator />).toJSON();
  expect(calculatorComponent).toMatchSnapshot();
});

test('expect calculator page to match working snapshot', () => {
  const calculatorPage = renderer.create(<CalcPage />).toJSON();
  expect(calculatorPage).toMatchSnapshot();
});

test('expect home page to match working snapshot', () => {
  const homePage = renderer.create(<Home />).toJSON();
  expect(homePage).toMatchSnapshot();
});

test('expect one plus two to equal 3', () => {
  const { container } = render(<Calculator />);
  const display = container.querySelector('input');

  const buttonOne = screen.getByText(/1/);
  const buttonTwo = screen.getByText(/2/);
  const additionButton = screen.getByText('+');
  const equalsButton = screen.getByText(/=/);

  fireEvent.click(buttonOne);
  fireEvent.click(additionButton);
  fireEvent.click(buttonTwo);
  fireEvent.click(equalsButton);

  expect(display.value).toEqual('3');
});

test('expect four times fives to equal 20', () => {
  const { container } = render(<Calculator />);
  const display = container.querySelector('input');

  const buttonFour = screen.getByText(/4/);
  const buttonFive = screen.getByText(/5/);
  const timesButton = screen.getByText('x');
  const equalsButton = screen.getByText(/=/);

  fireEvent.click(buttonFour);
  fireEvent.click(timesButton);
  fireEvent.click(buttonFive);
  fireEvent.click(equalsButton);

  expect(display.value).toEqual('20');
});

test('expect 200 modulo 3 equal 2', () => {
  const { container } = render(<Calculator />);
  const display = container.querySelector('input');

  const buttonZero = screen.getByText(/0/);
  const buttonTwo = screen.getByText(/2/);
  const buttonThree = screen.getByText(/3/);

  const moduloButton = screen.getByText('%');
  const equalsButton = screen.getByText(/=/);

  fireEvent.click(buttonTwo);
  fireEvent.click(buttonZero);
  fireEvent.click(buttonZero);
  fireEvent.click(moduloButton);
  fireEvent.click(buttonThree);
  fireEvent.click(equalsButton);

  expect(display.value).toEqual('2');
});
