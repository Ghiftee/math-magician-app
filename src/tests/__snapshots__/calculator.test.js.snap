// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect calculator element to match working snapshot 1`] = `
<div
  id="calculator"
>
  <input
    className="text-right border-0 bg-gray text-white p-1 py-3 w-100"
    readOnly={true}
    value={0}
  />
  <div
    className="buttons d-grid bg-gray"
  >
    <button
      onClick={[Function]}
      type="button"
    >
      AC
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      +/-
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      %
    </button>
    <button
      className="bg-warning"
      onClick={[Function]}
      type="button"
    >
      รท
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      7
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      8
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      9
    </button>
    <button
      className="bg-warning"
      onClick={[Function]}
      type="button"
    >
      x
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      4
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      5
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      6
    </button>
    <button
      className="bg-warning"
      onClick={[Function]}
      type="button"
    >
      -
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      1
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      2
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      3
    </button>
    <button
      className="bg-warning"
      onClick={[Function]}
      type="button"
    >
      +
    </button>
    <button
      id="zero-button"
      onClick={[Function]}
      type="button"
    >
      0
    </button>
    <button
      onClick={[Function]}
      type="button"
    >
      .
    </button>
    <button
      className="bg-warning"
      onClick={[Function]}
      type="button"
    >
      =
    </button>
  </div>
</div>
`;

exports[`expect calculator page to match working snapshot 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "justifyContent": "space-between",
      "padding": "20px",
    }
  }
>
  <div>
    <h4>
      Let's do some math!
    </h4>
  </div>
  <div
    id="calculator"
  >
    <input
      className="text-right border-0 bg-gray text-white p-1 py-3 w-100"
      readOnly={true}
      value={0}
    />
    <div
      className="buttons d-grid bg-gray"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        AC
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        +/-
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        %
      </button>
      <button
        className="bg-warning"
        onClick={[Function]}
        type="button"
      >
        รท
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        7
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        8
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        9
      </button>
      <button
        className="bg-warning"
        onClick={[Function]}
        type="button"
      >
        x
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        4
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        5
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        6
      </button>
      <button
        className="bg-warning"
        onClick={[Function]}
        type="button"
      >
        -
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        1
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        2
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        3
      </button>
      <button
        className="bg-warning"
        onClick={[Function]}
        type="button"
      >
        +
      </button>
      <button
        id="zero-button"
        onClick={[Function]}
        type="button"
      >
        0
      </button>
      <button
        onClick={[Function]}
        type="button"
      >
        .
      </button>
      <button
        className="bg-warning"
        onClick={[Function]}
        type="button"
      >
        =
      </button>
    </div>
  </div>
</div>
`;

exports[`expect home page to match working snapshot 1`] = `
<div>
  <h2
    style={
      Object {
        "marginBottom": "20px",
        "marginTop": "50px",
      }
    }
  >
    Welcome to our Page!
  </h2>
  <p>
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis id mattis felis. Nam accumsan leo justo. In convallis tellus sit amet iaculis tincidunt. Ut quis porttitor ante. Maecenas nec suscipit ex. Vivamus non sapien nisl. Vestibulum nisi tortor, cursus at feugiat quis, consequat mattis enim. Maecenas blandit lobortis erat, eget elementum massa ultricies in. Curabitur vestibulum porttitor nisl, nec efficitur mauris lobortis a. Curabitur iaculis gravida pulvinar. Sed vel dictum metus, consectetur consectetur lorem. Maecenas tellus est, congue vitae tellus hendrerit, consequat mattis lacus. Proin a ex ac magna venenatis lacinia at a leo. Suspendisse luctus dapibus elit, sit amet varius ligula. Aliquam vitae lectus vitae massa fermentum bibendum a eu ligula.
    <br />
    <br />
    Lorem ipsum dolor sit amet, consectetur adipiscing elit. Duis id mattis felis. Nam accumsan leo justo. In convallis tellus sit amet iaculis tincidunt. Ut quis porttitor ante. Maecenas nec suscipit ex. Vivamus non sapien nisl. Vestibulum nisi tortor, cursus at feugiat quis, consequat mattis enim. Maecenas blandit lobortis erat, eget elementum massa ultricies in. Curabitur vestibulum porttitor nisl, nec efficitur mauris lobortis a. Curabitur iaculis gravida pulvinar. Sed vel dictum metus, consectetur consectetur lorem. Maecenas tellus est, congue vitae tellus hendrerit, consequat mattis lacus. Proin a ex ac magna venenatis lacinia at a leo. Suspendisse luctus dapibus elit, sit amet varius ligula. Aliquam vitae lectus vitae massa fermentum bibendum a eu ligula.
  </p>
</div>
`;
